### PR TITLE
Delete expired docs from IndexedDB when not in cms

### PR DESCRIPTION
**Pull request by DirkDevelopmentFreelance**

### CODE REVIEW

**Comment by DirkDevelopmentFreelance** [comment]
@ivanslabbert I feel like this PR should not yet or can not yet be merged, as we discussed that clearing drafts has some stuff that has to be done in the API first

**Review by laurensgroeneveld — CHANGES_REQUESTED** [review]
Quite a few comments, but most of it boils down to that I think the initialization is not happening in exactly the right order, and that the app still doesn't purge expired docs. Did you test it in the browser?

File: shared/src/db/database.ts [hunk 1]
```diff
@@ -60,21 +61,32 @@ export type queryOptions = {
...
+    private isCms: boolean | undefined = undefined;
 
     constructor() {
         super("luminary-db");
 
+        if (config.getCmsFlag() == undefined || config.getCmsFlag() == null) {
+            throw new Error(
+                "Unable to load database - the CMS flag has not been set. The CMS flag should be set before instantiating the database.",
+            );
+        }
```
[the remark this answers is not in the document]
Comment (DirkDevelopmentFreelance) [reply shared/src/db/database.ts:78]: Yes, doing it this way, will immediately notify the developer if there is a sequencing error. For example when someone else updates the config. This ensures that a flag has in fact been set. This is actually a mistake on my side. The Configuration should no longer take a default config(After writing this I just saw the comment you made on the default value(s)). This ensures that the init function has to be called so relevant configs to each "project" is set.

File: shared/src/db/database.ts [hunk 2]
```diff
@@ -60,21 +61,32 @@ export type queryOptions = {
...
+
+        this.isCms = config.getCmsFlag();
```
[the remark this answers is not in the document]
Comment (DirkDevelopmentFreelance) [reply shared/src/db/database.ts:80]: So far I have not picked up any errors, maybe after the default value(s) has been removed, there will be errors, I will be taking a look at this now.

File: shared/src/db/database.ts [hunk 3]
```diff
@@ -626,6 +638,30 @@ class database extends Dexie {
...
+                if (expiryDate && expiryDate <= now) return true;
+                return false;
+            });
+            if (expiredDocs.length > 0) {
+                console.info("There are expired Documents...deleting");
+                try {
+                    await this.docs.bulkDelete(expiredDocs.map((doc) => doc._id));
+                } catch (e) {
+                    console.error(`Error:   ${e}`);
+                }
```
[the remark this answers is not in the document]
Comment (DirkDevelopmentFreelance) [reply shared/src/db/database.ts:660]: I think this should be a topic in the daily standup?

File: shared/src/db/database.spec.ts [hunk 4]
```diff
@@ -830,4 +833,115 @@ describe("baseDatabase.ts", () => {
...
+                video: "",
+                publishDate: 1704114000000,
+                expiryDate: expiredDate,
+                image: "test-image.jpg",
+            },
+        ];
+
+        const mockDb = new database();
+
+        await mockDb.docs.clear();
```
[the remark this answers is not in the document]
Comment (DirkDevelopmentFreelance) [reply shared/src/db/database.spec.ts:936]: When developing this fixed the issue I was encountering

File: shared/src/luminary.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { type SharedConfigT, config } from "./config";
+
+export function init(init: SharedConfigT) {
+    config.setConfig({ cms: init.cms });
+}
```
[the remark this answers is not in the document]
Comment (DirkDevelopmentFreelance) [reply shared/src/luminary.ts]: You are correct, after making changes and removing the default value(s), this error takes place inside of the test.

File: shared/src/db/database.spec.ts [hunk 6]
```diff
@@ -830,4 +836,40 @@ describe("baseDatabase.ts", () => {
             });
         });
     });
+    // it("deletes expired documents when not in cms-mode", async () => {
```
Comment (laurensgroeneveld) [inline shared/src/db/database.spec.ts:839]: This commented out test is still here. I think it is important that we have a test for the delete expired functionality.

Perhaps one way would be to `new` up a `database`, but first override the `config.cms` value? Not sure if calling `initLuminaryShared` again in here would work, but maybe that's also a way to go.

**Review by laurensgroeneveld — COMMENTED** [review]
This looks good now, just need to add a test on the expired functionality

File: shared/src/db/database.spec.ts (hunk 6 above)
Comment (DirkDevelopmentFreelance) [reply shared/src/db/database.spec.ts]: Implemented

File: shared/src/db/database.spec.ts [hunk 7]
```diff
@@ -830,4 +837,40 @@ describe("baseDatabase.ts", () => {
...
+                expiryDate: futureExpiredDate,
+            },
+        ];
+
+        await db.docs.clear();
+        await db.docs.bulkPut(docs);
+
+        (await db.docs.toArray()).every((val) => console.log(val));
+
+        await (db as any).deleteExpired();
```
[the remark this answers is not in the document]
Comment (DirkDevelopmentFreelance) [reply shared/src/db/database.spec.ts:869]: Yeah. But I have noted it and will look into it. And if I see I can make changes in the future to make it better, or somebody else does then we can just improve it.

File: shared/src/db/database.spec.ts [hunk 8]
```diff
@@ -830,4 +837,40 @@ describe("baseDatabase.ts", () => {
...
+            {
+                ...mockEnglishContentDto,
+                expiryDate: futureExpiredDate,
+            },
+        ];
+
+        await db.docs.clear();
+        await db.docs.bulkPut(docs);
+
+        (await db.docs.toArray()).every((val) => console.log(val));
```
[the remark this answers is not in the document]
Comment (DirkDevelopmentFreelance) [reply shared/src/db/database.spec.ts]: Took out the leftover console.log().

**Pull request merged**